headers/Ex1.hpp

```diff
@@ -1,65 +1,72 @@
+#include <memory> //unique_ptr
+
 template <typename T>
 class DynamicBuffer {
 public:
     explicit DynamicBuffer(size_t size);
     DynamicBuffer(const DynamicBuffer<T>& other);  
+    DynamicBuffer(DynamicBuffer<T> &&other) noexcept;
+
     DynamicBuffer<T>& operator=(const DynamicBuffer<T>& other); 
-    ~DynamicBuffer() noexcept;
+    DynamicBuffer<T>& operator=(DynamicBuffer<T>&& other) noexcept; 
+
+    ~DynamicBuffer() noexcept = default;
 
 private:
-    T* Allocate(size_t size);
-    void DeAllocate() noexcept;
-    void CopyDynamicBuffer(DynamicBuffer<T>& src);
+    void CopyDynamicBuffer(const DynamicBuffer<T>& src);
 
     size_t m_size;
-    T* m_buffer;
+    std::unique_ptr<T[]> m_buffer;
 };
 
 /****************************************** Main Functions *****************************************/
 
-template <typename T>
-DynamicBuffer<T>::DynamicBuffer(size_t size): m_buffer(Allocate(size)), m_size(size) {}
+template <typename T> //Ctor
+DynamicBuffer<T>::DynamicBuffer(size_t size): m_buffer(std::make_unique<T[]>(size)), m_size(size) {}
 
-template <typename T>
-DynamicBuffer<T>::DynamicBuffer(const DynamicBuffer<T>& other): m_buffer(Allocate(other.m_size)), m_size(other.m_size) {
+template <typename T> //CCtor
+DynamicBuffer<T>::DynamicBuffer(const DynamicBuffer<T>& other): m_buffer(std::make_unique<T[]>(other.m_size)), m_size(other.m_size) {
 
     CopyDynamicBuffer(other);
 }
 
-template <typename T>
+template <typename T> //Move Ctor
+DynamicBuffer<T>::DynamicBuffer(DynamicBuffer<T>&& other) noexcept : m_buffer(std::move(other.m_buffer)), m_size(other.m_size) {
+
+    other.m_size = 0;
+}
+
+
+template <typename T> //Copy Assignment
 DynamicBuffer<T>& DynamicBuffer<T>::operator=(const DynamicBuffer<T>& other) {
 
     if (this != &other) {
-        DeAllocate();
+        m_buffer = std::make_unique<T[]>(other.m_size);
+        m_size = other.m_size;
+
         CopyDynamicBuffer(other);
     }
+    
     return *this;
 }
 
-template <typename T>
-DynamicBuffer<T>::~DynamicBuffer() noexcept {
-
-    DeAllocate();
-}
+template <typename T> //Move Assignment
+DynamicBuffer<T>& DynamicBuffer<T>::operator=(DynamicBuffer<T>&& other) noexcept {
 
-/****************************************** Helper Functions *****************************************/
+    if (this != &other) {
+        m_buffer = std::move(other.m_buffer);
+        m_size = other.m_size;
 
-template <typename T>
-T* DynamicBuffer<T>::Allocate(size_t size) {
+        other.m_size = 0;
+    }
 
-    return new T[size];
+    return *this;
 }
 
-template <typename T>
-void DynamicBuffer<T>::DeAllocate() noexcept {
-
-    m_size = 0;
-    delete[] m_buffer;
-    m_buffer = nullptr;
-}
+/****************************************** Helper Functions *****************************************/
 
 template <typename T>
-void DynamicBuffer<T>::CopyDynamicBuffer(DynamicBuffer<T>& src) {
+void DynamicBuffer<T>::CopyDynamicBuffer(const DynamicBuffer<T>& src) {
```

Comment:
No need to do something about it, just think about it - Would this compile for T which has only move assignment?